core/src/main/scala/org/apache/spark/util/Utils.scala

```diff
@@ -1986,6 +1986,115 @@ private[spark] object Utils extends Logging {
+
+    def startOnce(tryPort: Int): (Option[T], Int) = {
+      val serviceString = if (serviceName.isEmpty) "" else s" '$serviceName'"
+      try {
+        val (service, port) = startService(tryPort)
+        logInfo(s"Successfully started service$serviceString on port $port.")
+        return (Some(service), port)
+      } catch {
+        case e: Exception if isBindCollision(e) =>
+          logWarning(s"Service$serviceString could not bind on port $tryPort. ")
+          (None, -1)
+      }
+    }
+
+    def startFromAvailable(rand: Boolean = false): (T, Int) = {
+
+      val portsStr = conf.get("spark.mesos.executor.port.ranges")
+
+      if (Option(portsStr).isEmpty) {
+        // should never happen
+        return startServiceOnPort(startPort, startService, conf, serviceName)
+      }
+      val ports = portRangeToList(portsStr)
+
+      val filteredPorts = ports.map(r => (r._1 to r._2).toList).reduce(_ ++ _).
+        distinct.filterNot(_ == startPort)
+
+      val maxRetries = Math.min(portMaxRetries(conf), filteredPorts.size)
+
+      val availPorts = {
+        if (rand) {
+          scala.util.Random.shuffle(filteredPorts.take(maxRetries)).toArray
+        } else {
+          filteredPorts.sorted.toArray
+        }
+      }
+      val serviceString = if (serviceName.isEmpty) "" else s" '$serviceName'"
+
+      for (offset <- 0 to maxRetries) {
+        val tryPort = availPorts(offset).toInt
+        try {
+          val (service, port) = startService(tryPort)
+          logInfo(s"Successfully started service$serviceString on port $port.")
+          return (service, port)
+        } catch {
+          case e: Exception if isBindCollision(e) =>
+            if (offset >= maxRetries) {
+              val exceptionMessage =
+                s"${e.getMessage}: Service$serviceString failed after $maxRetries retries!"
+              val exception = new BindException(exceptionMessage)
+              // restore original stack trace
+              exception.setStackTrace(e.getStackTrace)
+              throw exception
+            }
+            logWarning(s"Service$serviceString could not bind on port $tryPort.")
+        }
+      }
+
+      // Should never happen
+      throw new SparkException(s"Failed to start service$serviceString on port $startPort")
+    }
+
+    require(startPort == 0 || (1024 <= startPort && startPort < 65536),
+      "startPort should be between 1024 and 65535 (inclusive), or 0 for a " +
+        "random free port within range.")
+
+    if (startPort != 0) {
+
+      val (service, port) = startOnce(startPort)
+
+      if (port != -1) {
+        (service.get, port)
+      } else {
+        // try other
+        startFromAvailable()
+      }
+    }
+    else {
+      // try random
+      startFromAvailable(true)
+    }
+  }
+
+  def portRangeToList(ranges: String): List[(Long, Long)] = {
+    ranges.split(":").map { r => val ret = r.tail.reverse.tail.reverse.split(",");
+      (ret(0).toLong, ret(1).toLong) }.toList
```

Comment:
Public methods should have a short doc explaining what they do (what is the input format, what is the output like).

The implementation is also pretty complex. First, it'd be good to indent it more idiomatically. Second, the sequence of `tail.reverse` is hard to understand. It seems to only drop the first and the last elements, right? How about `r.substring(1, r.length - 1)`?